File: src/main/scala/homework3/java/EmployeeSeniority.java
package homework3.java;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.*;
import java.util.function.Predicate;
import java.util.function.Supplier;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;

public class EmployeeSeniority {
    enum Seniority {
        JUNIOR(s -> s.getSalary() < 14000),
        MIDDLE(s -> s.getSalary() >= 14000 && s.getSalary() < 21000),
        SENIOR(s -> s.getSalary() >= 21000);

        private Predicate<Employee> predicate;

        Seniority(Predicate<Employee> predicate) {
            this.predicate = predicate;
        }

        public Predicate<Employee> getPredicate(){
            return predicate;
        }
    }

    @Data
    @AllArgsConstructor
    static class Pair {
        Seniority key;
        Employee value;
    }

    public static void main(String[] args) {
        Random rnd = new Random();
        int minSalary = 5000;
        int maxSalary = 30000;
        int EmploeeyNumber = 5;

// Generate List of Emploee for test
        List<Employee> employees = rnd.ints(minSalary, maxSalary).
                mapToObj(i -> new Employee(("Emploee" + (rnd.nextInt(9999) + 1000)), i)).
                limit(EmploeeyNumber).collect(Collectors.toList());

        List<Seniority> seniorities = List.of(Seniority.JUNIOR, Seniority.MIDDLE, Seniority.SENIOR);

// Multimap of Lists of Employees by Seniority
        getSeniorityMap(employees, seniorities).entrySet().
                forEach(System.out::println);

// Numbers of Employees by Seniority
        getEmployeesNumberMap(employees, seniorities).
                entrySet().forEach(System.out::println);

// Max salalry by Seniority
        getMaxSalaryMap(employees, seniorities).
                entrySet().forEach(System.out::println);

// Total salalry by Seniority
        getTotalSalaryMap(employees, seniorities).
                entrySet().forEach(System.out::println);
    }

    private static Map<Seniority, Integer> getTotalSalaryMap(List<Employee> employees, List<Seniority> seniorities) {
        return getSeniorityMap(employees, seniorities).
                entrySet().stream().
                collect(Collectors.toMap(Map.Entry::getKey,
                                            p -> p.getValue().stream().mapToInt(emp -> emp.getSalary()).sum()));
    }

    private static Map<Seniority, Integer> getMaxSalaryMap(List<Employee> employees, List<Seniority> seniorities) {
        return getSeniorityMap(employees, seniorities).
                entrySet().stream().
                collect(Collectors.toMap(Map.Entry::getKey,
                                            p -> p.getValue().stream().mapToInt(emp -> emp.getSalary()).max().orElse(0)));
    }

    private static Map<Seniority, Integer> getEmployeesNumberMap(List<Employee> employees, List<Seniority> seniorities) {
        return getSeniorityMap(employees, seniorities).
                entrySet().stream().
                collect(Collectors.toMap(Map.Entry::getKey, p -> p.getValue().size()));
    }

    private static Map<Seniority, List<Employee>> getSeniorityMap(List<Employee> employees, List<Seniority> seniorities) {
        return seniorities.stream().
                flatMap(s -> employees.stream().
                        filter(s.getPredicate()).
                        map(emp -> new Pair(s, emp))
                ).collect(groupingBy(pair -> pair.getKey(), getCollector()));
    }

    private static Collector<Pair, List<Employee>, List<Employee>> getCollector() {
            return Collector.of(
                    () -> new ArrayList(),
                    (list, pair) -> list.add(pair.getValue()),
                    (left, right) -> { left.addAll(right); return left;},
                    Collector.Characteristics.IDENTITY_FINISH);
    }


}

File: src/main/scala/homework3/scala/EmployeeSeniority.scala
package homework3.scala

import scala.util.Random

object EmployeeSeniority {
  val minSalary = 5000
  val maxSalary = 30000
  val EmploeeyNumber = 10

  object Seniority extends Enumeration {
    val JUNIOR, MIDDLE, SENIOR = Value
    type Seniority = Value
  }
  import Seniority._

  val employees = LazyList.range(0, EmploeeyNumber, 1).
    map(n => Employee("Employee"+Random.between(1000, 9999), Random.between(minSalary, maxSalary))).
    toList

  val seniorities: Array[(Seniority.Value, Employee => Boolean)] =
    Array(
      (JUNIOR, s => s.salary < 14000),
      (MIDDLE, s => s.salary >= 14000 && s.salary < 21000),
      (SENIOR, s => s.salary >= 21000)
    )

  def main(args: Array[String]): Unit = {
    seniorityMap.
      foreach(x => {print(x._1+" ["); x._2.foreach(y => print(y + ", ")); println("]")})

    employeesNumberMap.
      foreach(println)

    maxSalaryMap.
      foreach(println)

    totalSalaryMap.
      foreach(println)
  }

  def seniorityMap: Map[Value, Array[Employee]] = seniorities.
      flatMap(seniority => employees.filter(seniority._2).map(emp => (seniority._1, emp))).
      groupMap(tup => tup._1)(tup => tup._2)

  def employeesNumberMap: Map[Value, Int] = seniorityMap.map(tup => (tup._1, tup._2.length))

  def maxSalaryMap: Map[Value, Int] = seniorityMap.
      map(tup => (tup._1, tup._2.map(emp => emp.salary).max))

  def totalSalaryMap: Map[Value, Int] = seniorityMap.
      map(tup => (tup._1, tup._2.map(emp => emp.salary).sum))
}

case class Employee(name: String, salary: Int) {
}


